openmetrics/src/main/java/io/airlift/openmetrics/MetricsResource.java

```diff
@@ -0,0 +1,340 @@
+    {
+        if (objectName.getDomain().contains("_")) {
+            log.warn("Unable to expose JMX metric with domain name %s, package names with underscores are unsupported.", objectName.getDomain());
+            throw new RuntimeException("Bad domain name %s".formatted(objectName.getDomain()));
+        }
+
+        StringBuilder metricNameBuilder = new StringBuilder()
+                .append(objectName.getDomain().replace(".", "_"));
+
+        if (objectName.getKeyProperty("name") != null) {
+            metricNameBuilder.append(NAME_SEPARATOR)
+                    .append(objectName.getKeyProperty("name"));
+        }
+
+        if (objectName.getKeyProperty("type") != null) {
+            metricNameBuilder.append(TYPE_SEPARATOR)
+                    .append(objectName.getKeyProperty("type").replace("$", "_"));
+        }
+
+        metricNameBuilder.append(ATTRIBUTE_SEPARATOR)
+                .append(attributeName.replace(".", "_"));
+
+        String metricName = metricNameBuilder.toString();
+
+        if (!metricNameValid(metricName)) {
+            log.warn("Calculated metric name has invalid characters %s skipping", metricName);
+        }
+        return metricName;
+    }
+
+    private boolean metricNameValid(String metricName)
+    {
+        return VALID_NAMES_CACHE.getUnchecked(metricName);
+    }
+
+    private List<String> toMetricExposition(String metricName)
+    {
+        try {
+            String attributeName = attributeNameFromMetricName(metricName);
+            return objectNamesFromMetricName(metricName).stream()
+                    .map(objectName -> getMetric(objectName, attributeName, metricName, ""))
+                    .flatMap(Optional::stream)
+                    .map(Metric::getMetricExposition)
+                    .collect(toImmutableList());
+        }
+        catch (MalformedObjectNameException e) {
+            log.warn(e, "Unable to retrieve metric %s.", metricName);
+            return ImmutableList.of();
+        }
+    }
+
+    private Optional<Metric> getMetric(ObjectName objectName, String attributeName, String metricName, String description)
+    {
+        try {
+            Object attributeValue = mbeanServer.getAttribute(objectName, attributeName);
+
+            if (attributeValue == null) {
+                return Optional.empty();
+            }
+
+            if (attributeValue instanceof Number) {
+                return Optional.of(Gauge.from(metricName, (Number) attributeValue, description));
+            }
+
+            return Optional.empty();
+        }
+        catch (JMException ex) {
+            log.debug(ex, "Unable to get metric for ObjectName %s and Attribute %s.", objectName.getCanonicalName(), attributeName);
+            return Optional.empty();
+        }
+    }
+
+    private String inferAttributesForObjectName(ObjectName objectName)
+    {
+        StringBuilder expositions = new StringBuilder();
```

Comment:
There are `Metric` to `String` conversions and then `String` concatenations at multiple layers in this metrics traversal logic. 
E.g we concatenate here in `inferAttributesForObjectName` and the method which calls it (`jmxMetricExpositions`).

I think the code would be easier to read if we obtained all the `Metric` objects we care about and we had just one place when we call `getMetricExposition` and concatenate those to build the response (in `getMetrics`).

Obtaining list of `Metric` objects can be done with an accumulator (E.g sth based on `ImmutableList.Builder`) you pass around and add stuff to it. The changed method signatures would look sth like:
* `void managedMetricExpositions(Consumer<Metric>)`
* `void getMetricsRecusively(String prefix, ManagedClass managedClass, Consumer<Metric> accumulator)`
* `String jmxMetricExpositions(ObjectName initialObjectName)` -> `void jmxMetrics(ObjectName initialObjectName, Consumer<Metric> accumulator)`
* ...

I believe code-wise it would very good chage. But maybe there is a reason why we cannot do so? E.g. maybe is keeping `Metric` objects in memory for request duration significantly more costly than theirs `String` based expositions?